EIPS/eip-4804.md

```diff
@@ -0,0 +1,140 @@
+   Calldata: 0x
+   ...
+}
+```
+
+## Motivation
+
+Currently, reading data from Web3 generally relies on a translation done by a Web2 proxy to Web3 blockchain. The translation is mostly done by the proxies such as dApp websites/node service provider/etherscan, which are out of the control of users. The standard here aims to provide a simple way for Web2 users to directly access the content of Web3.
+
+## Specification
+
+The specification here only specifies for the read EVM message. Write message will be defined in the future proposals.
+
+A Web3 URL is in the following form
+
+```
+Web3URL = "web3://" [userinfo "@"] contractName [":" chainid] ["->(" returnTypes ")"] path [? query]
+contractName = address | name "." nsProvider
+path = ["/" method ["/" argument_0 ["/" argument_1 ... ]]]
+argument = [type "!"] value
+```
+
+where
+
+- "web3://" indicates the Web3 URL **schema**.
+- **userinfo** indicates which user is calling the EVM, i.e., "From" field in EVM call message. If not specified, the protocol will use 0x0 as the sender address.
+- **contractName** indicates the contract to be called, i.e., "To" field in the EVM call message. If the **contractName** is an **address**, i.e., 0x + 20-byte-data hex, then "To" will be the address. Otherwise, the name is from a name service. In the second case, **nsProvider** will be the short name of name service providers such as "ens", "w3q", etc. The way to translate the name from a name service to an address will be defined in future proposals.
+- **chainid** indicates which chain to call the message. If not specified, the protocol will use the same chain as the name service provider, e.g., 1 for eth, and 333 for w3q. If no name service provider is available, the default chainid is 1.
+- **returnTypes** tells the format of the returned data. If not specified, the returned message data will be parsed in "(bytes32)" and MIME will be set based on the suffix of the last argument. If **returnTypes** is "()", the returned data will be parsed in raw bytes in JSON.  Otherwise, the returned message will be parsed in the specified "returnTypes" in JSON.  
+
+### Resolve Mode
+
+Once the "To" address and chainid are determined, the protocol will check the resolver mode of contract by calling "resolveMode" method. The protocol currently supports two resolve modes:
+
+#### Manual Mode
+
+The manual mode will not do any interpretation of **path**, and put **path** as calldata of the message directly.
+
+#### Auto Mode
+
+The auto mode is the default mode of resolver (also applies when the "resolveMode" method is unavailable in the target contract). In the auto mode, if **path** is empty, then the protocol will call the target contract with empty calldata. Otherwise, the calldata of the EVM message will use standard Ethereum contract ABI [Contract ABI Specification — Solidity 0.8.3 documentation](https://docs.soliditylang.org/en/v0.8.3/abi-spec.html), where
```

Comment:
Please avoid including external links.

```suggestion
The auto mode is the default mode of resolver (also applies when the "resolveMode" method is unavailable in the target contract). In the auto mode, if **path** is empty, then the protocol will call the target contract with empty calldata. Otherwise, the calldata of the EVM message will use standard Solidity contract ABI, where
```